Doc/library/asyncio-task.rst

```diff
@@ -861,19 +861,46 @@ Waiting Primitives
 
 .. function:: as_completed(aws, *, timeout=None)
 
-   Run :ref:`awaitable objects <asyncio-awaitables>` in the *aws*
-   iterable concurrently.  Return an iterator of coroutines.
-   Each coroutine returned can be awaited to get the earliest next
-   result from the iterable of the remaining awaitables.
+   Run :ref:`awaitable objects <asyncio-awaitables>` in the *aws* iterable
+   concurrently. The returned object can be iterated to obtain earliest next
+   results of the awaitables as they finish.
+
+   The object returned by ``as_completed()`` can be iterated as an
+   :term:`asynchronous iterator` or a plain :term:`iterator`. When asynchronous
+   iteration is used, the originally-supplied awaitables are yielded if they
+   are tasks or futures. This makes it easy to correlate previously-scheduled
+   tasks with their results. Example::
+
+       ipv4_connect = create_task(
+           open_connection("127.0.0.1", 80)
+       )
+       ipv6_connect = create_task(
+           open_connection("::1", 80)
+       )
+       tasks = [ipv4_connect, ipv6_connect]
+
+       async for earliest_connect in as_completed(tasks):
+           reader, writer = await earliest_connect
+           if earliest_connect is ipv6_connect:
+               print('IPv6 connection established.')
+           else:
+               print('IPv4 connection established.')
+           # ...
 
-   Raises :exc:`TimeoutError` if the timeout occurs before
-   all Futures are done.
+   During asynchronous iteration, implicitly-created tasks will be yielded for
+   supplied awaitables that aren't tasks or futures.
 
-   Example::
+   When used as a plain iterator, each iteration yields a new coroutine that
+   returns the result or raises the exception of the next completed awaitable.
+   This pattern is compatible with Python versions older than 3.13::
 
-       for coro in as_completed(aws):
-           earliest_result = await coro
-           # ...
+        for coro in as_completed(aws):
+            earliest_result = await coro
+            # ...
```

Comment:
It would be misleading to name the synchronous `for` variable `earliest_connect`, because it is not one of `ipv4_connect` or `ipv6_connect`. If you expand `# ...` to the same code as in the asynchronous iteration example, it will not fail, but work incorrectly.

Reply:
And yet, these two are completely equivalent:

```py
    async for result in as_completed([ipv4_connect, ipv6_connect]):
        r, w = await result
```

and

```py
    for result in as_completed([ipv4_connect, ipv6_connect]):
        r, w = await result
```

It's true that `result` represents something different, but in most cases you just want to await it, and then it's the same.

Maybe we need two examples -- one showing that if you just await the result, there's no difference; another showing that only `async for` can identify the task that first completed?

Reply:
True. And note that in the asynchronous iteration example it could not necessary be `await`, it could be a synchronous operation, like `result()` or `result.get()`. `await` is only used to mimic the old code.

Reply:
Oh, the result is already 'done' in that case? I had not realized that subtlety. Worth pointing out in the docs then!